### PR TITLE
Support DN to string conversion #154

### PR DESCRIPTION
**Pull request by asambeka**
Added String method to dn.go, for supporting DN to string conversion.

### CODE REVIEW

File: dn.go [hunk 1]
```diff
@@ -208,6 +208,27 @@ func (d *DN) AncestorOf(other *DN) bool {
...
+
+	for i := range d.RDNs {
+		if i > 0 {
+			buffer.WriteString(",")
+		}
+		for j := range d.RDNs[i].Attributes {
+			if j > 0 {
+				buffer.WriteString("+")
+			}
+			buffer.WriteString(d.RDNs[i].Attributes[j].Type)
```
Comment (liggitt) [inline dn.go]: where is escaping done?

Comment (asambeka) [reply dn.go]: Value inspection and escaping (RFC 2253, 4514) is not implemented.  Only canonical representation with spaces removed from [=,+] was done. I can add value inspection and escaping.

Comment (asambeka) [reply dn.go]: @liggitt I have added the changes for handing escaping of attribute values as per RFC 4512, 4514. Implementation is generic at DN level and can be used easily.

File: dn.go [hunk 2]
```diff
@@ -208,6 +209,81 @@ func (d *DN) AncestorOf(other *DN) bool {
...
+			buffer.WriteString("=")
+			//Escape the value before building DN string
+			val := EscapeAttrValue(d.RDNs[i].Attributes[j].Value)
```
Comment (asambeka) [inline dn.go]: @liggitt Here is the change (EscapeAttrValue), please review.

Comment (asambeka) [reply dn.go]: @liggitt Can you please provide your review inputs on this? - Thnx

**Comment by asambeka** [comment]
@liggitt Please check and approve the PR

**Comment by johnweldon** [comment]
@asambeka - I really appreciate your work on this.  I have a couple reservations that incline me to want to use #104 instead of this PR:

1. The other PR adds the encoding to the `AttributeTypeAndValue` type directly rather than in a loop on the containing type. I prefer keeping the encoding close to the type that actually is affected by it.
1. This PR uses regular expressions, rather than the simple parse loop in the other PR.  I feel that regular expression use for this is overkill and less user friendly for understanding what the code is doing.
1. The other PR handles encoding hex values, which this PR neglects to do.

I am sorry I didn't review the differences more thoroughly before you worked on the updates today.  Have I missed any benefits of this PR over the other one that I did not enumerate?

**Comment by johnweldon** [comment]
Feel free to reopen or make a new PR if you'd like to pursue this further.  Thanks for you contribution.

**Pull request closed without merging**